scene/gui/dialogs.cpp

```diff
@@ -330,21 +341,24 @@ void AcceptDialog::remove_button(Control *p_button) {
 	ERR_FAIL_COND_MSG(button->get_parent() != buttons_hbox, vformat("Cannot remove button %s as it does not belong to this dialog.", button->get_name()));
 	ERR_FAIL_COND_MSG(button == ok_button, "Cannot remove dialog's OK button.");
 
+	Control *right_spacer = Object::cast_to<Control>(button->get_meta("__right_spacer"));
+	if (right_spacer) {
+		buttons_hbox->remove_child(right_spacer);
+		ERR_FAIL_COND_MSG(right_spacer->get_parent() != nullptr, vformat("Failed to remove button's spacer %s.", button));
+		button->remove_meta("__right_spacer");
+		memdelete(right_spacer);
+	}
+
+	buttons_hbox->remove_child(button);
+	ERR_FAIL_COND_MSG(button->get_parent() != nullptr, vformat("Failed to remove button %s.", button));
+
 	if (button->is_connected("pressed", callable_mp(this, &AcceptDialog::_custom_action))) {
 		button->disconnect("pressed", callable_mp(this, &AcceptDialog::_custom_action));
 	}
 	if (button->is_connected("pressed", callable_mp(this, &AcceptDialog::_cancel_pressed))) {
 		button->disconnect("pressed", callable_mp(this, &AcceptDialog::_cancel_pressed));
 	}
 
-	Node *right_spacer = buttons_hbox->get_child(button->get_index() + 1);
-	// Should always be valid but let's avoid crashing.
-	if (right_spacer) {
-		buttons_hbox->remove_child(right_spacer);
-		memdelete(right_spacer);
-	}
-	buttons_hbox->remove_child(button);
-
```

Comment:
@YuriSizov In #65446 you've changed the ordering of disconnecting/removing parts, any reason for that? :thinking:

I've changed it back so nothing is disconnected in case removing fails.

Reply:
Probably so everything is done in a neat order. Why do we expect the removal to fail? It looks very unusual to check after the removal if node's parent is `nullptr`. I don't think we do anything like that anywhere.

The only reason I can imagine is that the button, or the spacer, were moved elsewhere. But we can check that before trying to remove each node and fail early instead. In fact, for the `button` we already do that check.

Reply:
It could fail e.g. if trying to `remove_button` from not a main thread, or when the parent is "blocked" which would be the case e.g. if it's called when handling some notification for such button (it's propagated through the parent). In such cases `remove_child` would fail and error out, but if it would be followed by e.g. `memdelete(right_spacer)` it would likely result in a crash (or at least in an invalid internal state).

Yeah, there's the question where to draw the line between safe handling situations where user breaks things which they shouldn't change in the first place. And these I'd ignore. But `remove_child` can fail for kinda valid usages as mentioned above, I think these should be handled gracefully.

Reply:
I think the implications of being called when blocked or not on the main thread affect quite a lot of our codebase. And we mostly ignore those situations (you'll get an error, things will be kinda broken). So I wouldn't go out of our way to handle it here. But it is something we should discuss and devise a common solution for.

> but if it would be followed by e.g. memdelete(right_spacer) it would likely result in a crash (or at least in an invalid internal state).

We can `queue_free()` it, it should avoid the crash.

Reply:
You can keep the code as is, by the way. It's not making it worse. It's just not something that is consistent with the rest of the GUI codebase, and we still need to think about solving this properly in a more generic way. For example, this method is explicitly for removing children, so it can be locked the same way `remove_child` is. Through some helper perhaps. Though that is definitely beyond the scope here.

Reply:
Went ahead and already got rid of these. Also moved the disconnecting/removing back, so less changes overall. :upside_down_face: Yeah, it's problematic to keep everything being handled the same way across the codebase.